### DISCUSSION TITLE
Poll: StreamField block previews - developer experience

### DISCUSSION THREAD

**Discussion by thibaudcolas** [wagtail/wagtail#12674] [Polls]
@laymonage and I are working on support for [StreamField block previews](https://github.com/wagtail/wagtail/issues/11997) in the editor interface. That’s an improvement on our roadmap for the February 2025 release of Wagtail. We could use some feedback!

Here’s what the feature would look like for CMS users – a separate pane to the side of the existing block chooser:

<img width="994" alt="streamfield block previews screenshot of ui prototype" src="https://github.com/user-attachments/assets/f54d7f66-4601-428f-997b-1c711a734fcc">

For developers – there are lots of ways in which we could define those previews – and they all would take time to configure in block definitions. **Which of those preview options do you think would be a good fit for your projects?**

1. A `preview_image`. Path to a static image file, like a screenshot of the blocks as displayed on the site or as designed.
2. A `preview_template`. Path to a Django or Jinja template, that would then render a preview of the block with mock data.
3. A `template_component` using our existing [Template component pattern](https://docs.wagtail.org/en/stable/extending/template_components.html).
4. A preview mixin (similar to `PreviewableMixin` snippets and pages).

---

All options are valid, and if you wanted to consider multiple ones do comment to say so. Fundamentally this will be an opt-in feature, so we want to get a sense of:

1. Which approaches will see initial real-world adoption (a screenshot is easy to take, creating preview templates/components takes a bit of time)
2. Which approaches will be sustainable long-term (screenshots never "break" but they get out of date, rendered HTML is harder to get out of date but likely to break)

**Comment by thibaudcolas** [comment]
[Comment from](https://github.com/wagtail/wagtail/issues/11997#issuecomment-2522996478) @stevejalim:

> A selfish +1 from me for `preview_template`, because our FE templates are jinja (or rather non-core-Django Jinja via`django-jinja`).
>
> That said, because `Component.render_html` is overrideable, we can call anything we like to render, so there's also a +1 for me on `template_component` 
>
> It's a -1 on `preview_image` because that creates a level of developer complexity that would slow us down

**Comment by ababic** [comment]
I've got to go with the mixin class, because I feel like there'd be limits to what you could do in a template alone (e.g. You may want to pass a specific Image into the context to render a thumbnail)

**Comment by Nigel2392** [comment]
if going the mixin route:

possibly add media.Media object for the preview page? I can imagine some use cases where this would come in nice

the preview image might be able to get auto- generated, maybe with a management command, no?
or maybe this can be done on- hover of the block in the streamfield menu, and then be cached?
having an override to the auto- generation would be very nice though

might be doable with something like https://html2canvas.hertzen.com/ (html2canvas would probably require client- side logic for caching and such)


as for what @ababic said, probably yes. a lot of my blocks use a custom context, please do not limit us mortal developers to templates alone!

**Reply by thibaudcolas** [reply]
We’ve considered automated image generation a bit, feels very doable but would probably have to keep this as out of scope for Wagtail core. As a package though it’d make a lot of sense.

**Comment by enzedonline** [comment]
Choice of preview template or mixin would be ideal. Would also be good to be able to turn off the preview panel in the parameters when calling the streamfield/streamblock and/or in the Wagtail user profile. Maybe a sticky button on the block menu ... I can see this might be distracting for people already familiar with the blocks. 

On that, it would also be good to be able to turn off the block search on streambock instances since it's often redundant when there are only a handful of options to choose from, and to make the column count responsive (5x8 is a better experience than 2x20).

**Reply by Nigel2392** [reply]
> Choice of preview template or mixin would be ideal. Would also be good to be able to turn off the preview panel in the parameters when calling the streamfield/streamblock and/or in the Wagtail user profile. Maybe a sticky button on the block menu ... I can see this might be distracting for people already familiar with the blocks. 
> 
> 
> 
> On that, it would also be good to turn off the block search on streambock instances since it's often redundant when there are only a handful of options to choose from, and to make the column count responsive (5x8 is a better experience than 2x20).

On the block search - probably should be a PR but i think the search is pretty neat. it should not be removed; you are right however that it is often reduntant. probably this should be a check, or passed from meta options, like search_enabled=True or in django settings like WAGTAIL_STREAMBLOCK_SEARCH_ITEMS_COUNT = number to delimit at which block count search should be visible or not in the menu

**Comment by enzedonline** [comment]
I was thinking just as a parameter when initialising the StreamField or StreamBlock, something like `search=True` by default rather than adding a global setting.

```python
    body = StreamField(
        SimpleStreamBlock(), verbose_name="Page body", blank=True, search=False
    )
 ```
 and
```python
class CalloutBlock(StructBlock):
    heading = CalloutHeadingStructBlock(required=False, label=" ")
    content = CalloutStreamBlock(label=" ", search=False)
```
I have several nested StreamBlocks with just 2 or 3 block types - the search takes more space than the blocks themselves. At the moment, I hide it via CSS (and set the block list to responsive grid to use more of the screen width).

EDIT - ah, I'd meant to word that "it would also be good _to be able to_ turn off the block search"

**Reply by thibaudcolas** [reply]
Is it really a problem that it takes space if you only have 2 or 3 block types anyway? The search and filtering is also there to speed up keyboad-only interaction.

**Comment by seb-b** [comment]
+1 for a preview template. But I think something that would lower our overhead even greater would be a `get_preview_context` or similar method where we could set up dummy data and use the blocks _real_ template to render the preview, means we wouldn't have to keep two templates in sync when we make changes to a block. Most of our blocks are struct blocks with pretty easy to mock data, I realise this isn't the same case for everyone though, especially with nested blocks. If we're using the block's real template to render it means the preview data shouldn't really ever get stale or break, if it does then the real template is broken anyway. 

If the image route is chosen there's no way we would adopt this for our 30+ sites already in production, the overhead would be too much, and I can't imagine adopting it in new projects knowing the upkeep required

**Reply by thibaudcolas** [reply]
Is there a case that you could reuse the same images for the bulk of the 30 sites? I imagine they all have different designs but the fundamental concept of what the blocks are might be similar enough to warrant similar representations?

**Reply by seb-b** [reply]
They'd all be styled differently depending on the site, none of our sites really share common styles. I'd be very hesitant to put a preview for block that didn't look almost exactly like the end result, I'd be trying to avoid confusion for the user

**Reply by thibaudcolas** [reply]
👍  makes sense! But then you do assume a developer working on the block will remember to update its "preview mode" to be a truthful reflection of the end result. Which to me feels equally tricky for blocks that are anything beyond the simplest "single text field" ones?

**Reply by seb-b** [reply]
That goes to my comment of using the real template instead of a separate template for the preview. If we're updating the block, we're updating the template, we're (hopefully) updating the preview dummy data. 

We'd probably set up some unit tests to make sure the block renders with it's dummy data without breaking. 

But for more complex blocks I don't have a nice solution, If it's a complicated block and the maintenance of the preview becomes a burden I reckon we'd simply opt-out

**Comment by thibaudcolas** [comment]
Well this seems pretty clear! Thank you all 😌 . The people who voted for "template" – for blocks that already have their own template, would you expect to be reusing it or to instead set up something that’s using separate markup?

**Reply by stevejalim** [reply]
I wonder if that needs to be decided up front - maybe not? It's possible I'm not looking at this right, but if the block has a `ui_preview_template_name` attribute (or some better name), then it can be left to the developer to point to whatever file they want to use for the preview. 

Regardless of template use, the block could still automagically generate fake/[fuzzy](https://factoryboy.readthedocs.io/en/stable/fuzzy.html) data based on what the expected fields are (and ideally create a generic placeholder image where needed, such as in the style of https://placehold.co/)

HTH!

**Reply by thibaudcolas** [reply]
Yep, I think we’d try to leave it undecided – but most likely in the real world it’d have to be a separate template. I’m not sure if it’s clear any "live-rendered" block preview will depend on dedicated mock data and CSS. Most likely HTML markup around the block itself (at least enough so the preview iframe is accessible). Possibly JavaScript.

I’m team "preview image" because of all of this. Capturing screenshots once a year that will get outdated but would never break feels much simpler.

**Comment by thibaudcolas** [comment]
It’s on! #12700
